### ISSUE TITLE
Investigate and Improve application configuration and initialization (Admin; Python)

### ISSUE THREAD

**Issue by ccostino** [GSA/notifications-admin#701]
We've been making several improvements to the app and updating dependencies quite a bit recently, and it's time to review how the app is currently configured and initialized.  There are a few reasons for this:

- The app config as it stands is quite complicated and cumbersome and could at least benefit from some cleanup
- Newer versions of dependencies may have different and improved ways of getting things setup
   - In some cases, there may be required changes
   - In others, there might be simpler ways of accomplishing what we need now by leveraging the features of the libraries/modules
- There are other dependencies that we're not currently able to update yet (e.g., SQLAlchemy, gunicorn) until we know for sure our app is ready for them

**Note:** This is intended for the Python side of things for now.  Separate issues will be written for the JS side once we're ready.

## Implementation Sketch and Acceptance Criteria

- [x] Make sure all dependencies that we can easily update are up-to-date
- [x] Review the documentation of Flask, SQLAlchemy, and associated plugins and integrations for recommended application configuration and initialization
- [x] Update our application config with an eye toward reducing complexity and cleaning things up as much as possible; reduce moving parts as much as possible
- [x] Update related tests and/or write new tests as necessary to test the application config
- [x] Update our documentation to reflect any changes related to environment setup, local development setup, etc.

## Security Considerations

- Updating dependencies will make sure we stay ahead of any vulnerability disclosures and help us with app maintenance
- Any changes to sensitive data management with newer module updates will have to be accounted for
- Any documentation updates that touch on sensitive information will have to be handled with care

**Comment by ccostino** [comment]
@stvnrlly and @tdlowden, I've started [a document](https://docs.google.com/document/d/1mlLBdlQ6qCQJ6sZfdBQNcWq9sdGGUQnyLEPOXdiM8hk/edit#heading=h.rdex7n5whyrv) to capture what I'm finding in my research and observations so far that I'll keep adding to.  Some things are colored red to denote potential action items to take.

Once I've finished we can talk through some of these and that'll help put together a list of recommendations and formal tasks that we can create and assign.

I started with the admin site first because it's a similar architecture to the API but with a few less moving parts, so it's easier to figure out a new foundation to move toward from here and then build on it further with the additional components in the API.

**Comment by ccostino** [comment]
I've had several conversations with folks about Python dependency management and what might be taking place with the Cloud Foundry Python buildpack for dependency management and installation - full notes are written up in the document, but in summary the main takeaways are these:

- Leave our configuration as is for the app deployment; in fact, this is what we'll likely need to continue doing in the future (meaning generating the `requirements.txt` file ourselves as a deployment step).
- We should investigate switching over to using the `pyproject.toml` file as our source of truth for Python dependency management, and find a suitable tool to work with it (e.g., [`poetry`](https://python-poetry.org/) or [`pip-tools`](https://pip-tools.readthedocs.io/en/latest/).
- Make sure [Dependabot is configured correctly](https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#configuration-options-for-the-dependabotyml-file) for whichever tool we use.

Lastly, we'll want to keep these [secure installation practices](https://pip.pypa.io/en/latest/topics/secure-installs/) in mind with our overall configuration as well.

**Comment by stvnrlly** [comment]
> We should investigate switching over to using the `pyproject.toml` file

Here's a proposed ADR for `poetry`: https://github.com/GSA/notifications-api/issues/452

**Comment by ccostino** [comment]
@stvnrlly and I reviewed the things I've uncovered in the notes doc and decided upon the following:

- Move forward with the [`poetry` ADR](https://github.com/GSA/notifications-api/issues/452) and shifting to using `pyproject.toml` as separate work for Python dependency management (note that we'll still need to generate `requirements.txt` for deployment)
- Write up an ADR for reorganizing and making more use of [Flask Blueprints](https://flask.palletsprojects.com/en/2.3.x/blueprints/) for better code organization
- Write up another ADR for splitting the main app config (and in particular the error handling) in separate modules to help with code organization and maintenance
- Write up a separate issue exploring the configuration and use of [`gunicorn`](https://docs.gunicorn.org/en/latest/index.html) and if there are any other adjustments we can or should make (noting that it'll have to account for operating in the cloud.gov environment).

Overall, there's nothing of an obvious nature that needs to be addressed immediately.  These are more incremental improvements that we'll have to work toward over time.

Once I create these new issues, ADRs, etc., we'll close this issue out.
